### PR TITLE
Social: Add pricing table to Jetpack Social

### PR DESCRIPTION
**Pull request by danielpost**
#### Changes proposed in this Pull Request:

* Add a pricing table to Jetpack Social for sites that have the share limit enabled.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

1202647675190352-as-1202849147697167/f

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

* Create a new site, connect to Jetpack and make sure the Jetpack Social admin page doesn't show a pricing table.
* Enable share limits (D85993-code)
* Reload the Jetpack Social admin page and check that there's a pricing table present.
* Choose "Start for free" and confirm that the table disappears.
* Reload the page and confirm the table is still gone.
* Run `wp option update jetpack-social_show_pricing_page true`.
* Reload the page again, but this time choose the paid plan.
* Go through checkout, and confirm:
  * The pricing table is gone again.
  * The share limits meter is gone, and the cards are shown instead.

### CODE REVIEW

File: projects/plugins/social/src/class-rest-controller.php [hunk 1]
```diff
@@ -62,15 +62,58 @@ public function require_admin_privilege_callback() {
...
+				'key'                 => 'show_pricing_page',
+				'get_callback'        => function () {
+					return Jetpack_Social::should_show_pricing_page();
+				},
+				'update_callback'     => function ( $value ) {
+					return update_option( Jetpack_Social::JETPACK_SOCIAL_SHOW_PRICING_PAGE_OPTION, (int) $value );
+				},
+				'validation_callback' => '__return_true',
+			),
+		);
```
Comment (danielpost) [inline projects/plugins/social/src/class-rest-controller.php]: This is intended to be an easy way to control what settings we return in our endpoint, and how they get updated.

I had a nagging feeling I was recreating core functionality but couldn't find anything. Happy to be proven wrong though!

File: projects/plugins/social/src/js/store/reducer/shares-data.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+const sharesData = ( state = {} ) => {
+	return state;
+};
+
+export default sharesData;
```
Comment (danielpost) [inline projects/plugins/social/src/js/store/reducer/shares-data.js:5]: This was missing in an earlier PR—fixed here.

File: projects/plugins/social/src/class-jetpack-social.php [hunk 3]
```diff
@@ -159,9 +160,11 @@ public function initial_state() {
 				'apiRoot'           => esc_url_raw( rest_url() ),
 				'apiNonce'          => wp_create_nonce( 'wp_rest' ),
 				'registrationNonce' => wp_create_nonce( 'jetpack-registration-nonce' ),
+				'siteSuffix'        => ( new Status() )->get_site_suffix(),
```
Comment (danielpost) [inline projects/plugins/social/src/class-jetpack-social.php:191]: We can update this when https://github.com/Automattic/jetpack/pull/26135 gets merged.

File: projects/plugins/social/src/class-rest-controller.php (hunk 1 above)
Comment (pablinos) [reply projects/plugins/social/src/class-rest-controller.php]: I think you might be able to handle some of this by defining the schema for the endpoint. [The argument schema](https://developer.wordpress.org/rest-api/extending-the-rest-api/schema/#argument-schema) should allow validation and sanitisation for the parameters.

File: projects/plugins/social/src/class-rest-controller.php [hunk 4]
```diff
@@ -80,17 +123,32 @@ public function get_settings() {
...
 		// If we're not changing anything, just return the current settings.
-		if ( $publicize_active === null ) {
+		if ( empty( $settings ) ) {
 			return rest_ensure_response( $this->get_settings() );
 		}
 
-		$publicize_active_updated = ( new Modules() )->update_status( \Jetpack_Social::JETPACK_PUBLICIZE_MODULE_SLUG, $publicize_active, false, false );
+		$errors = array();
+
+		foreach ( $settings as $setting ) {
```
Comment (pablinos) [inline projects/plugins/social/src/class-rest-controller.php]: I think we might be able to simplify this again, by looping over the fields provided, once the schema is defined.

File: projects/plugins/social/src/class-rest-controller.php [hunk 5]
```diff
@@ -62,15 +62,58 @@ public function require_admin_privilege_callback() {
...
+
 	/**
 	 * GET `jetpack/v4/social/settings`
 	 */
 	public function get_settings() {
-		return rest_ensure_response(
-			array(
-				'publicize_active' => ( new Modules() )->is_active( \Jetpack_Social::JETPACK_PUBLICIZE_MODULE_SLUG ),
-			)
+		$settings = $this->get_available_settings();
```
Comment (pablinos) [inline projects/plugins/social/src/class-rest-controller.php]: With the schema defined, we should be able to loop through the requested fields. We could still use the call backs though.

File: projects/plugins/social/src/class-rest-controller.php [hunk 6]
```diff
@@ -62,15 +62,58 @@ public function require_admin_privilege_callback() {
...
+	 * - validation_callback: Gets the return value of the update_callback.
+	 *     Return a boolean to indicate if the setting updated successfully.
+	 *
+	 * @return array
+	 */
+	public function get_available_settings() {
+		return array(
+			array(
+				'key'                 => 'publicize_active',
+				'get_callback'        => function () {
```
Comment (pablinos) [inline projects/plugins/social/src/class-rest-controller.php]: I notice this is repeated in [the main Jetpack_Social class](https://github.com/Automattic/jetpack/pull/26213/files#diff-d2a85dafd23b905630dba0fd25897b63c873bfea0d0e83b9a3ce237f1dbde996R166). Perhaps we should define it as a named function somewhere.

File: projects/plugins/social/src/class-rest-controller.php [hunk 7]
```diff
@@ -62,15 +62,58 @@ public function require_admin_privilege_callback() {
...
+				},
+				'validation_callback' => function ( $return ) {
+					return ! is_wp_error( $return );
+				},
+			),
+			array(
+				'key'                 => 'show_pricing_page',
+				'get_callback'        => function () {
+					return Jetpack_Social::should_show_pricing_page();
+				},
```
Comment (pablinos) [inline projects/plugins/social/src/class-rest-controller.php]: As this is a callable, could it be?

```suggestion
				'get_callback'        => array( 'Jetpack_Social', 'should_show_pricing_page' ),
```

If we carry on using callbacks, the others could be defined in a similar way.

File: projects/plugins/social/src/class-jetpack-social.php (hunk 3 above)
Comment (pablinos) [reply projects/plugins/social/src/class-jetpack-social.php:191]: That PR ended up not adding it to the initial state, so we'll probably need to keep that here. I feel like we should be able to combine this state with the editor initial state and use `getSiteFragment` for both, but that's a much bigger change!

File: projects/plugins/social/src/js/components/admin-page/index.jsx [hunk 8]
```diff
@@ -36,6 +48,20 @@ const Admin = () => {
 		);
 	}
 
+	if ( isShareLimitEnabled && ! hasPaidPlan && showPricingPage ) {
+		return (
+			<AdminPage moduleName={ __( 'Jetpack Social 1.0', 'jetpack-social' ) } header={ <Logo /> }>
```
Comment (pablinos) [inline projects/plugins/social/src/js/components/admin-page/index.jsx]: Do you think we could prevent repeating this? The contents of the one below could be wrapped in a fragment perhaps.

Should we take this opportunity to make the version number dynamic? We could define the `moduleName` as a const earlier on.

**Review by pablinos — COMMENTED** [review]
It's looking good so far. I like the idea of the callbacks in the rest controller, but we might be able to simplify things a bit with a schema.

File: projects/plugins/social/src/js/components/admin-page/index.jsx (hunk 8 above)
Comment (danielpost) [reply projects/plugins/social/src/js/components/admin-page/index.jsx]: Done in [5181979](https://github.com/Automattic/jetpack/pull/26213/commits/5181979cf81db3adf903f4828aa7d67098a2fd7c).

File: projects/plugins/social/src/class-rest-controller.php (hunk 6 above)
Comment (danielpost) [reply projects/plugins/social/src/class-rest-controller.php]: Done in [390d332](https://github.com/Automattic/jetpack/pull/26213/commits/390d3321d5f8e2970b8c2295675fa341f64def62).

**Review by pablinos — APPROVED** [review]
Looks good, let's merge it!

**Pull request merged**